### PR TITLE
fix fixtures on multiple databases

### PR DESCRIPTION
**Pull request by MichaelSp**
When loading fixtures in multiple databases the integrity check is only disabled on one connection (the default). This patch disables the integrity check on all connections and starts a transaction on all connections to load the fixtures into all databases. Finally it restores the integrity check.

As far as I know there is no github issue for this, yet.
I'd like to see this in 4.2 but I can also port this to the 5.0 branch.

Unfortunately I can't provide tests, as I had some issues to have a multi-database test set up running. Can someone more experienced elaborate on this?

PR for 4.2 branch: https://github.com/rails/rails/pull/24952

### CODE REVIEW

**Comment by MichaelSp** [comment]
No decision yet?
And again: I'd still like to have this in 4.2 as well! (https://github.com/rails/rails/pull/24952)

**Comment by rafaelfranca** [comment]
Could you try the master branch? I believe this PR should fix it https://github.com/rails/rails/pull/20818. And sorry we don't backport this kind of structural change in stable releases so it is going to be only in 5.1.

**Comment by MichaelSp** [comment]
Sorry, I'm locked on 4.x. Cant test/verify this in 5.x.
I understand that you don't want to support #20818 in 4.x but  #24952 isn't a big thing. And after all: it's a bug!

**Comment by arthurnn** [comment]
I will take a look at it this week, sorry for the wait.

**Comment by MichaelSp** [comment]
looong week @arthurnn  :grin:

File: activerecord/lib/active_record/fixtures.rb [hunk 1]
```diff
@@ -520,53 +520,64 @@ def self.create_fixtures(fixtures_directory, fixture_set_names, class_names = {}
...
 
+    # disables referential integrity on all connections and
+    # starts a transaction on all connection. Finally it executes the
+    # given block.
+    def self.disable_referential_integrity_in_transaction connections, &block
+      return block.call if connections.empty?
+      connection = connections.shift
+      connection.disable_referential_integrity do
+        connection.transaction(:requires_new => true) do
+          disable_referential_integrity_in_transaction connections, &block
```
Comment (arthurnn) [inline activerecord/lib/active_record/fixtures.rb]: Why this is a recursive method?

File: activerecord/lib/active_record/fixtures.rb [hunk 2]
```diff
@@ -520,53 +520,64 @@ def self.create_fixtures(fixtures_directory, fixture_set_names, class_names = {}
...
 
-          update_all_loaded_fixtures fixtures_map
+        update_all_loaded_fixtures fixtures_map
 
-          connection.transaction(:requires_new => true) do
-            deleted_tables = Set.new
-            fixture_sets.each do |fs|
-              conn = fs.model_class.respond_to?(:connection) ? fs.model_class.connection : connection
-              table_rows = fs.table_rows
+        all_connections = ActiveRecord::Base.connection_handler.connection_pool_list.map(&:connection)
```
Comment (arthurnn) [inline activerecord/lib/active_record/fixtures.rb]: Do we need to start a transaction in all connections from the pool? That seems too much, I think just the first connection(current) of each pool should be enough. 
What I mean by too much is, the test thread will hold reference to all connections, not allowing any new thread to checkout a connection from the pool.

**Review by arthurnn — COMMENTED** [review]
We need something like this. 
Left some comments, and also to move forward with this we need some tests for a functionality like that.

File: activerecord/lib/active_record/fixtures.rb (hunk 2 above)
Comment (matthewd) [inline activerecord/lib/active_record/fixtures.rb]: Should we instead be gathering `klass.connection` during the earlier loop?

File: activerecord/lib/active_record/fixtures.rb (hunk 1 above)
Comment (MichaelSp) [inline activerecord/lib/active_record/fixtures.rb]: This is the only way (I could come up with) do `disable_referential_integrity` AND `transaction` using the block syntax. 
Iterative approach using `each` wouldn't work here.

I think it's easier to read/understand than that pseudo code:
```
connections.each do |connection|
begin
  connection.disable_referential_integrity
  connection.start_transaction
  yield
  connection.check_if_rollback ? connection.rollback : connection.commit
  connection.enable_referential_integrity
rescue SomeExceptions
ensure
  other stuff
end
```

**Comment by MichaelSp** [comment]
Again, it doesn't make sense for me to elaborate this in the rails 5 as I'm unfortunately locked on 4.x. I need this in rails 4. So if there is any chance to get this merged in 4, I can try to come up with tests for that (4 and 5).

File: activerecord/lib/active_record/fixtures.rb [hunk 3]
```diff
@@ -520,53 +520,64 @@ def self.create_fixtures(fixtures_directory, fixture_set_names, class_names = {}
...
+        disable_referential_integrity_in_transaction(all_connections) do
+          deleted_tables = Set.new
+          fixture_sets.each do |fs|
+            conn = fs.model_class.respond_to?(:connection) ? fs.model_class.connection : connection
```
Comment (arthurnn) [inline activerecord/lib/active_record/fixtures.rb]: What happens if User connects to connection A, and we are in a new transaction on connection B?

File: activerecord/lib/active_record/fixtures.rb (hunk 1 above)
Comment (arthurnn) [inline activerecord/lib/active_record/fixtures.rb]: yeah, make sense

**Review by arthurnn — COMMENTED** [review]
@matthewd any extra thoughts?

**Pull request closed without merging**